haystack/modeling/model/biadaptive_model.py

```diff
@@ -485,3 +485,67 @@ def convert_from_transformers(
             bi_adaptive_model.connect_heads_with_processor(processor.tasks)  # type: ignore
 
         return bi_adaptive_model
+
+
+class BiAdaptiveSharedModel(BiAdaptiveModel):
```

Comment:
As far as I understand this, you are using the same language model as an encoder. So instead of the original DPR architecture where we have a question and a document encoder, you are using the same encoder for both.

I'd argue that this is the same architecture as used by sentence transformers (https://sbert.net/docs/training/overview.html) and maybe we should consider if it makes sense to re-implement all of the functionality already provided there or if it makes sense to use sentence transformers under the hood.

Reply:
> As far as I understand this, you are using the same language model as an encoder. So instead of the original DPR architecture where we have a question and a document encoder, you are using the same encoder for both.

Exactly. But, most importantly, MDR also uses a different retrieval mechanism that iteratively adds context documents (see https://ar5iv.labs.arxiv.org/html/2009.12756#S2.SS2.SSS0.Px1).

> I'd argue that this is the same architecture as used by sentence transformers (https://sbert.net/docs/training/overview.html) and maybe we should consider if it makes sense to re-implement all of the functionality already provided there or if it makes sense to use sentence transformers under the hood.

Yes, I agree that it's quite similar to sentence transformers, except for the multi-hop part, which they don't support AFAIK.

For now, I based my implementation on Haystack's `DensePassageRetriever`, which uses Hugging Face's [DPR](https://huggingface.co/docs/transformers/v4.19.2/en/model_doc/dpr), and then calls the document store's `query_by_embedding`. It took only a few changes to adapt this for MDR. I'm not sure if there is much to be gained when using sentence transformers. What you see in my changes now is all there was to do. But please let me know if you find parts that can be simplified by using sentence transformers 😊